src/stan_math_backend/Transform_Mir.ml

```diff
@@ -30,6 +30,24 @@ let rec change_kwrds_stmts s =
     | x -> map Fn.id change_kwrds_stmts x in
   {s with pattern}
 
+let eval_eigen_indexed e =
+  let open Expr.Fixed in
+  let f ({pattern; meta} as expr) =
+    match (pattern, UnsizedType.is_eigen_type (Expr.Typed.type_of expr)) with
+    | Indexed _, true ->
+        {meta; pattern= FunApp (StanLib ("eval", FnPlain, AoS), [expr])}
+    | _ -> expr in
```

Comment:
`Indexed` is not enough. Some math lib functions also fail.
```stan
 vector test2(vector gamma) {
   int D = num_elements(gamma);

   if (D == 1)
      return rep_vector(D, 0);
    else
      return test2(head(gamma,D - 1));
  }
```

Reply:
It's easy enough to also wrap arbitrary math functions, but figuring out an exhaustive list of which ones will trigger this behavior seems daunting. It's probably reasonable to start with the functions listed [here](https://mc-stan.org/docs/functions-reference/slicing-and-blocking-functions.html)? If we get reports or find more in the future it is easy enough to add them